README.md
README
======

This is just a repackaged version of the tclodbc package done by:

    Roy Nurmi (main author)
    Roy.Nurmi@iki.fi
    
    Tom Poindexter (Unix port)
    tpoindex@nyx.net
    
    Rob Savoye (Unix port)
    rob@cygnus.com

The current configure and Makefile.in were updated to the TEA 2002
specification by Jeff Hobbs.

Please see the license.txt file for more information about the licensing of
this package.

SUPPORTED
---------
Supported (tested) platforms:
    - Visual C++ 6.0 on Windows NT
    - So far I have not done any Unix testing...

    
BINARY DISTRIBUTION
-------------------
### Windows NT
For Windows NT, you will probably install the included binary
distribution. To do this, just type the following (tclsh may be
tclsh82 or something similar):

c:\temp> tclsh setup.tcl
Installation successful.

The resulting installation may be different than one done with the
Makefile. The setup.tcl file does not know about --prefix and
--exec-prefix, so it installs into the DLL into the standard Tcl package
directory.

### Unix
There is no Unix binary distribution.


BUILDING
--------
### Windows NT

Follow the steps from the Unix section. If you want to rebuild the package
without cygwin under Windows NT, the included tclodbc.mak can be used to
build the library.  You will have to adjust the project to work properly on
your configuration.

### Unix

Under Unix you will have to use autoconf to create a configuration file, and
thus Makefile, which will build and install the package.  Specify the same
--prefix and --exec-prefix that you used to build Tcl.  You may also have to
specify --with-tcl to point to the directory that contains the tclConfig.sh
file.

On Ubuntu 14.04, the CC and LIBS environment variables must be used to
correctly link the right libraries.

    export CC=g++
    export LIBS=-lodbcinst

To configure and make:

    autoconf
    ./configure --prefix=/usr/local/tcl --exec-prefix=/usr/local/tcl/bin
    make
    make install

If your ODBC include files and libraries can not be found
automatically, you may want to use the following two options:

    --with-odbcinclude, directory where ODBC headers are
    --with-odbclibrary, directory where ODBC libraries are


USING THE PACKAGE
-----------------
The extension is installed as a tcl package. After installation the
extension is loaded simply by command:

    package require tclodbc

This command adds the command "database" to the tcl command
interpreter (or actually loads the package on demand). Command
"database" is used for creating database connection objects as
well as for configuring the system data sources.


####################
The original README.unix follows here:

Tclodbc on Unix

I have made a preliminary configure and Makefile for Tclodbc.

The only IODBC drivers I have were the free ones. The one I use is now
included in the PostgreSQL 6.4 release. This was developed under
RedHat Linux 5.1, using G++ (cygnus-2.91.47).

  --with-tclinclude       directory where tcl headers are
  --with-tclconfig        directory containing tcl configuration (tclConfig.sh)
  --with-tkinclude        directory where tk headers are
  --with-tkconfig         directory containing tk configuration (tkConfig.sh)
  --with-odbcinclude      directory where ODBc headers are
  --with-odbclibrary      directory where ODBC headers are

Also, configure expects to find a Tcl installation in the paths specified by
--prefix= and --exec-prefix=.  You should specify the same directories as
when you configured Tcl. Chances are you won't need to specify any
configuration options. An example where the ODBC files are in a non
standard place is:

	 mkdir obj
	 cd obj
	 /usr/local/src/tclodbc/configure \
		--with-odbcinclude=/usr/local/include/iodbc \
		--with-odbclibrary=/usr/local/lib/iodbc

You may also need to tweak the Makefile to set the correct C++
compiler option flag to build a shared library. This uses the flags as
determined by Tcl to build a shared library by default. As long as you
use the same compiler to build Tcl and Tclodbc, everything should work
fine.

'make install' copies the library to your Tcl library, makes a directory for
Tclodbc with version number, and builds a pkgIndex.tcl file.

If all goes right, all you need to use it is:

	$ tclsh8.0 (or cygtclsh8.0 when using cygwin)
	% package require Tclodbc
	2.1
	% database datasources
	{....

Rob Savoye
rob@cygnus.com

#################
Tclodbc on Unix

I have made a preliminary configure and Makefile for Tclodbc.

The only ODBC drivers to which I have access on Unix is Redbrick's, so I've
left the ODBC configuration fairly simple.  The configure script expects:

	--with-odbc-directory=DIR
	--with-odbc-library=LIBSPEC

The Makefile uses the ODBC directory to prefix $ODBC_DIR/include and 
$ODBC_DIR/lib to find include and libraries.  You may have to tweak the
Makefile after configuration if your ODBC installation is different.
The ODBC library specification is just the library names you would use to
link, eg. '-lodbc'

Also, configure expects to find a Tcl installation in the paths specified by
--prefix= and --exec-prefix=.  You should specify the same directories as
when you configured Tcl.

Example:
	./configure --prefix=/home/tpoindex --with-odbc-directory=/redbrick \
				--with-odbc-library=-lrbodbc

Configure uses the Autoconf macros to find your C++ compiler.

'make'
The Makefile will build a shared library for Tclodbc.  You may also need to
tweak the Makefile to set the correct C++ compiler option flag to build a 
shared library.  See the define for TCL_SHLIB_LD.

'make install' copies the library to your Tcl library, makes a directory for
Tclodbc with version number, and builds a pkgIndex.tcl file.

If all goes right, all you need to use it is:

	$ tclsh8.0
	% package require Tclodbc
	1.2
	% database datasources
	{....


My development environment is:
	Dec Unix (unam -a: OSF1 xxx.xxx.com V4.0 564 alpha)
	Redbrick ODBC driver for Dec (version: V5.0.12)
	Dec C++ compiler (cxx -V: DEC C++ V5.3-004 on Digital UNIX (Alpha))

Tom Poindexter
tpoindex@nyx.net


ChangeLog
2006-01-25  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure, configure.in: update to TEA 3.5
	* tclconfig/tcl.m4:        TEA rcs 1.89 2006/01/25 21:25:02

2005-03-25  Jeff Hobbs  <jeffh@ActiveState.com>

	* tclconfig/tcl.m4: TEA 3.2 updates
	* configure, configure.in, Makefile.in: OS X patch from Steffen
	* ODBC.m4, tclodbc.hxx: iodbc checks improved

2004-03-20  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure, configure.in, setup.tcl: update to v2.5
	* database.cxx, statemnt.cxx, strings.cxx, tclodbc.hxx:
	* doc/reference.htm (id): add optional list of lists result sets
	with 'db set multisets true' and '$stmt moreresults'. (riesterer)

2004-03-19  Jeff Hobbs  <jeffh@ActiveState.com>

	* encoding.cxx, statemnt.cxx, database.cxx, tclodbc.hxx:
	correct assignment warning, casts and whitespace.

	* tclodbc.cxx, strings.cxx: TCLODBC_VERSION -> PACKAGE_VERSION
	* ODBC.m4, configure.in, configure, Makefile.in, tclconfig/tcl.m4:
	updated to the latest TEA base

2003-03-13  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure:
	* tclconfig/tcl.m4: updated to latest TEA sample

2002-05-01  Jeff Hobbs  <jeffh@ActiveState.com>

	* tclodbc.hxx (new):
	* tclodbc.hxx.in (removed): removed need for input file.  Moved
	tcl.h to being included near the beginning.

	* setup.tcl: code cleanup

	* README:
	* README.cygwin: updated notes about new TEA setup

	* Makefile.in:      Replaced old configure setup with TEA 2002
	* ODBC.m4:          specification.  Full update of configure.in,
	* aclocal.m4:       ODBC.m4 and Makefile.in.  Verified on Win2K
	* configure (new):  and Linux-2.4 with UnixODBC.
	* configure.in:
	* config/install-sh (removed):
	* config/mkinstalldirs (removed):
	* config/tcl.m4 (removed):
	* mkIndex.tcl.in (removed):
	* tclconfig/README.txt (new):
	* tclconfig/install-sh (new):
	* tclconfig/tcl.m4 (new):

	Created ChangeLog file post-2.3 tclodbc release


encoding.cxx
/**********************************************************************

                  - TCLODBC COPYRIGHT NOTICE - 

This software is copyrighted by Roy Nurmi, contact address by email at
Roy.Nurmi@iki.fi. The following terms apply to all files associated with 
the software unless explicitly disclaimed in individual files.

The author hereby grant permission to use, copy, modify, distribute,
and license this software and its documentation for any purpose, provided
that existing copyright notices are retained in all copies and that this
notice is included verbatim in any distributions. No written agreement,
license, or royalty fee is required for any of the authorized uses.
Modifications to this software may be copyrighted by their authors
and need not follow the licensing terms described here, provided that
the new terms are clearly indicated on the first page of each file where
they apply.

IN NO EVENT SHALL THE AUTHORS OR DISTRIBUTORS BE LIABLE TO ANY PARTY
FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
ARISING OUT OF THE USE OF THIS SOFTWARE, ITS DOCUMENTATION, OR ANY
DERIVATIVES THEREOF, EVEN IF THE AUTHORS HAVE BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.

THE AUTHORS AND DISTRIBUTORS SPECIFICALLY DISCLAIM ANY WARRANTIES,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.  THIS SOFTWARE
IS PROVIDED ON AN "AS IS" BASIS, AND THE AUTHORS AND DISTRIBUTORS HAVE
NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR
MODIFICATIONS.

GOVERNMENT USE: If you are acquiring this software on behalf of the
U.S. government, the Government shall have only "Restricted Rights"
in the software and related documentation as defined in the Federal 
Acquisition Regulations (FARs) in Clause 52.227.19 (c) (2).  If you
are acquiring the software on behalf of the Department of Defense, the
software shall be classified as "Commercial Computer Software" and the
Government shall have only "Restricted Rights" as defined in Clause
252.227-7013 (c) (1) of DFARs.  Notwithstanding the foregoing, the

  authors grant the U.S. Government and others acting in its behalf
permission to use and distribute the software in accordance with the
terms specified in this license. 

***********************************************************************/

#include <string.h>
#include <tcl.h>

#if TCL_MAJOR_VERSION == 8 && TCL_MINOR_VERSION >= 1

void FreeTclodbcEncodedString(Tcl_Obj *p) {
    //extern Tcl_ObjType TclodbcEncodedString;
    Tcl_DString* pds = (Tcl_DString*) p->internalRep.twoPtrValue.ptr2;
    if (pds) {
        Tcl_DStringFree(pds);
        Tcl_Free((char*)pds);
    }
    if (p->internalRep.twoPtrValue.ptr1)
        Tcl_FreeEncoding((Tcl_Encoding) p->internalRep.twoPtrValue.ptr1);
    p->internalRep.twoPtrValue.ptr1 = NULL;
    p->internalRep.twoPtrValue.ptr2 = NULL;
    p->typePtr = NULL;
}

void DuplicateTclodbcEncodedString(Tcl_Obj *src, Tcl_Obj *dst) {
    Tcl_DString* pds1;
    Tcl_DString* pds2;

    // copy encoding value
    dst->internalRep.twoPtrValue.ptr1 = src->internalRep.twoPtrValue.ptr1;

    // duplicate DString
    pds1 = (Tcl_DString*) src->internalRep.twoPtrValue.ptr2;
    if (pds1) {
        pds2 = (Tcl_DString*) Tcl_Alloc(sizeof(Tcl_DString));
        Tcl_DStringInit(pds2);
        Tcl_DStringAppend(pds2,  Tcl_DStringValue(pds1), Tcl_DStringLength(pds1));
        dst->internalRep.twoPtrValue.ptr2 = pds2;
    } else {
        dst->internalRep.twoPtrValue.ptr2 = NULL;
    }
}

void UpdateTclodbcEncodedString(Tcl_Obj *o) {
    unsigned int len;
    Tcl_Encoding e = (Tcl_Encoding) o->internalRep.twoPtrValue.ptr1;
    Tcl_DString* pds = (Tcl_DString*) o->internalRep.twoPtrValue.ptr2;

    // This is needed only, if external value is actually stored in ptr2.
    // In any other case the representations are equal, and decoding is
    // not necessary.
    if (pds) {
        // this encodes strings from 'normal' strings to UTF-8, which is
        // the internal representation of tcl 8.0
        Tcl_DString ds;
        Tcl_DStringInit(&ds);
        Tcl_ExternalToUtfDString (e, Tcl_DStringValue(pds), Tcl_DStringLength(pds), &ds);

        len = Tcl_DStringLength(&ds);
        o->bytes = Tcl_Alloc(len+1);
        o->bytes[len] = '\0';
        o->length = len;
        memcpy (o->bytes, Tcl_DStringValue(&ds), len);
        Tcl_DStringFree(&ds);
    }
}

int SetTclodbcEncodedString(Tcl_Interp *interp, Tcl_Obj *p) {
    if (interp)
        Tcl_SetResult(interp, (char *)"Not supported", TCL_STATIC);

    return TCL_ERROR;
}

Tcl_ObjType TclodbcEncodedString = {
    (char *)"TclodbcEncodedString", 
    FreeTclodbcEncodedString, 
    DuplicateTclodbcEncodedString, 
    UpdateTclodbcEncodedString, 
    SetTclodbcEncodedString
};

#endif


strings.cxx
/**********************************************************************

                  - TCLODBC COPYRIGHT NOTICE - 

This software is copyrighted by Roy Nurmi, contact address by email at
Roy.Nurmi@iki.fi. The following terms apply to all files associated with 
the software unless explicitly disclaimed in individual files.

The author hereby grant permission to use, copy, modify, distribute,
and license this software and its documentation for any purpose, provided
that existing copyright notices are retained in all copies and that this
notice is included verbatim in any distributions. No written agreement,
license, or royalty fee is required for any of the authorized uses.
Modifications to this software may be copyrighted by their authors
and need not follow the licensing terms described here, provided that
the new terms are clearly indicated on the first page of each file where
they apply.

IN NO EVENT SHALL THE AUTHORS OR DISTRIBUTORS BE LIABLE TO ANY PARTY
FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
ARISING OUT OF THE USE OF THIS SOFTWARE, ITS DOCUMENTATION, OR ANY
DERIVATIVES THEREOF, EVEN IF THE AUTHORS HAVE BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.

THE AUTHORS AND DISTRIBUTORS SPECIFICALLY DISCLAIM ANY WARRANTIES,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.  THIS SOFTWARE
IS PROVIDED ON AN "AS IS" BASIS, AND THE AUTHORS AND DISTRIBUTORS HAVE
NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR
MODIFICATIONS.

GOVERNMENT USE: If you are acquiring this software on behalf of the
U.S. government, the Government shall have only "Restricted Rights"
in the software and related documentation as defined in the Federal 
Acquisition Regulations (FARs) in Clause 52.227.19 (c) (2).  If you
are acquiring the software on behalf of the Department of Defense, the
software shall be classified as "Commercial Computer Software" and the
Government shall have only "Restricted Rights" as defined in Clause
252.227-7013 (c) (1) of DFARs.  Notwithstanding the foregoing, the
authors grant the U.S. Government and others acting in its behalf
permission to use and distribute the software in accordance with the
terms specified in this license. 

***********************************************************************/

//////////////////////////////////////////////////////////////////////////
// STRINGS AND STRINGTABLES
//

#include "tclodbc.hxx"

// standard strings
const char* strVersion = "tclodbc " PACKAGE_VERSION " (c) Roy Nurmi 1998, 2002";
const char* strMemoryAllocationFailed = "Memory allocation failed";
const char* strInvalidHandle		= "Invalid handle";
const char* strOK			= "OK";
const char* strTables			= "tables";
const char* strColumns		= "columns";
const char* strIndexes		= "indexes";
const char* strPrimarykeys		= "primarykeys";
const char* strTypeinfo		= "typeinfo";
const char* strEval			= "eval";
const char* strRead			= "read";
const char* strGet			= "get";
const char* strSet			= "set";
const char* strAutocommit		= "autocommit";
const char* strCommit			= "commit";
const char* strRollback		= "rollback";
const char* strWrongArgs		= "wrong # args";
const char* strUsage = 
"Usage:\n"
"  database [connect] id datasourcename userid password\n"
"  database [connect] id connectionstring\n"
"  database configure operation driver attributelist\n"
"  database datasources\n"
"  database drivers\n"
"  database version\n";
const char* strCmdNotAvailable	= "command not available";
const char* strEmpty			= "";
const char* strDisconnect		= "disconnect";
const char* strRun			= "run";
const char* strDrop			= "drop";
const char* strExecute		= "execute";
const char* strMoreResults		= "moreresults";
const char* strFetch			= "fetch";

const char* strConcurrency		= "concurrency";
const char* strMaxrows		= "maxrows";
const char* strTimeout		= "timeout";
const char* strMaxlenght		= "maxlenght"; // stupid spelling mistake before 2.2, retained for compatibility
const char* strMaxlength		= "maxlength";
const char* strRowsetsize		= "rowsetsize";
const char* strCursortype		= "cursortype";
const char* strNoscan			= "noscan";
const char* strEncoding		= "encoding";
const char* strStatement		= "statement";
const char* strRowCount		= "rowcount";
const char* strAsyncEnable		= "async";

const char* strInvalidOption		= "Invalid option: ";

const char* strOldSyntax = "This is old syntax which has been removed from tclodbc 2.1. Check documentation for replacement";

// IMPORTANT NOTE

// in tclodbc 2.1 the NumStr structure is optimized for
// speed with binary search. This requires that the strings
// in the struct are given in alphabetical order.
// The first row in a table gives the total count of strings
// in the struct.

NumStr sqlType [] = {
    {19,		NULL	       },
    {SQL_BIGINT,	(char *) "BIGINT"       },
    {SQL_BINARY,	(char *) "BINARY"       },
    {SQL_BIT,		(char *) "BIT"	        },
    {SQL_CHAR,		(char *) "CHAR"	        },
    {SQL_DATE,		(char *) "DATE"	        },
    {SQL_DECIMAL,	(char *) "DECIMAL"      },
    {SQL_DOUBLE,	(char *) "DOUBLE"       },
    {SQL_FLOAT,		(char *) "FLOAT"	},
    {SQL_INTEGER,	(char *) "INTEGER"      },
    {SQL_LONGVARBINARY,	(char *) "LONGVARBINARY"},
    {SQL_LONGVARCHAR,	(char *) "LONGVARCHAR"  },
    {SQL_NUMERIC,	(char *) "NUMERIC"      },
    {SQL_REAL,		(char *) "REAL"	        },
    {SQL_SMALLINT,	(char *) "SMALLINT"     },
    {SQL_TIME,		(char *) "TIME"	        },
    {SQL_TIMESTAMP,	(char *) "TIMESTAMP"    },
    {SQL_TINYINT,	(char *) "TINYINT"      },
    {SQL_VARBINARY,	(char *) "VARBINARY"    },
    {SQL_VARCHAR,	(char *) "VARCHAR"      }
};

NumStr attrDef [] = {
    {12,			NULL		},
    {SQL_COLUMN_DISPLAY_SIZE,	(char *) "displaysize"	},
    {SQL_COLUMN_LABEL,		(char *) "label"	},
    {SQL_COLUMN_NAME,		(char *) "name"		},
    {SQL_COLUMN_NULLABLE,	(char *) "nullable"	},
    {SQL_COLUMN_OWNER_NAME,	(char *) "owner"	},
    {SQL_COLUMN_PRECISION,	(char *) "precision"	},
    {SQL_COLUMN_QUALIFIER_NAME,	(char *) "qualifiername"},
    {SQL_COLUMN_SCALE,		(char *) "scale"	},
    {SQL_COLUMN_TABLE_NAME,	(char *) "tablename"	},
    {SQL_COLUMN_TYPE,		(char *) "type"		},
    {SQL_COLUMN_TYPE_NAME,	(char *) "typename"	},
    {SQL_COLUMN_UPDATABLE,	(char *) "updatable"	},
};

NumStr stmtOp [] = {
    {9,				NULL		},
    {SQL_ASYNC_ENABLE,		strAsyncEnable	},
    {SQL_CONCURRENCY,		strConcurrency	},
    {SQL_CURSOR_TYPE,		strCursortype	},
    {SQL_MAX_LENGTH,		strMaxlenght	}, // incorrectly spelled here, retained for compatibility
    {SQL_MAX_LENGTH,		strMaxlength	},
    {SQL_MAX_ROWS,		strMaxrows	},
    {SQL_NOSCAN,		strNoscan	},
    {SQL_ROWSET_SIZE,		strRowsetsize	},
    {SQL_QUERY_TIMEOUT,		strTimeout	}
};

NumStr configOp [] = {
#ifdef INCLUDE_EXTENSIONS
    {6,			       NULL	       },
    {ODBC_ADD_DSN,	       (char *) "add_dsn"       },
    {ODBC_ADD_SYS_DSN,	       (char *) "add_sys_dsn"   },
    {ODBC_CONFIG_DSN,	       (char *) "config_dsn"    },
    {ODBC_CONFIG_SYS_DSN,      (char *) "config_sys_dsn"},
    {ODBC_REMOVE_DSN,	       (char *) "remove_dsn"    },
    {ODBC_REMOVE_SYS_DSN,      (char *) "remove_sys_dsn"}
#else
    {0,			       NULL	       }
#endif
};

NumStr connectOp [] = {
    {11,			NULL		},
    {SQL_ASYNC_ENABLE,		strAsyncEnable	},
    {SQL_AUTOCOMMIT,		strAutocommit	},
    {SQL_CONCURRENCY,		strConcurrency	},
    {SQL_CURSOR_TYPE,		strCursortype	},
    {TCLODBC_ENCODING,		strEncoding	},
    {SQL_MAX_LENGTH,		strMaxlenght	}, // incorrectly spelled here, retained for compatibility
    {SQL_MAX_LENGTH,		strMaxlength	},
    {SQL_MAX_ROWS,		strMaxrows	},
    {SQL_NOSCAN,		strNoscan	},
    {SQL_ROWSET_SIZE,		strRowsetsize	},
    {SQL_QUERY_TIMEOUT,		strTimeout	}
};

NumStr booleanOp [] = {
    {8,				NULL		},
    {0,				(char *) "0"		},
    {1,				(char *) "1"		},
    {0,				(char *) "false"	},
    {0,				(char *) "no"		},
    {0,				(char *) "off"		},
    {1,				(char *) "on"		},
    {1,				(char *) "true"		},
    {1,				(char *) "yes"		}
};

NumStr concurrencyOp [] = {
    {4,				NULL	     },
    {SQL_CONCUR_LOCK,		(char *) "lock"	     },
    {SQL_CONCUR_READ_ONLY,	(char *) "readonly"  },
    {SQL_CONCUR_ROWVER,		(char *) "rowver"    },
    {SQL_CONCUR_VALUES,		(char *) "values"    }
};

NumStr cursorOp [] = {
    {4,				NULL		},
    {SQL_CURSOR_DYNAMIC,	(char *) "dynamic"	},
    {SQL_CURSOR_FORWARD_ONLY,	(char *) "forwardonly"	},
    {SQL_CURSOR_KEYSET_DRIVEN,	(char *) "keysetdriven"	},
    {SQL_CURSOR_STATIC,		(char *) "static"	}
};

NumStr databaseOptions [] = {
    {5,				NULL	     },
    {TclDatabase::CONFIGURE,	(char *) "configure"   	},
    {TclDatabase::CONNECT,	(char *) "connect"     	},
    {TclDatabase::DATASOURCES,	(char *) "datasources" 	},
    {TclDatabase::DRIVERS,	(char *) "drivers"	},
    {TclDatabase::VERSION,	(char *) "version"	}
};

NumStr databaseCmds [] = {
    {15,			NULL	      },
    {TclDatabase::AUTOCOMMIT,	strAutocommit },
    {TclDatabase::COLUMNS,	strColumns    },
    {TclDatabase::COMMIT,	strCommit     },
    {TclDatabase::DISCONNECT,	strDisconnect },
    {TclDatabase::EVAL,		strEval	      },
    {TclDatabase::GET,		strGet	      },
    {TclDatabase::INDEXES,	strIndexes    },
    {TclDatabase::PRIMARYKEYS,	strPrimarykeys},
    {TclDatabase::READ,		strRead	      },
    {TclDatabase::ROLLBACK,	strRollback   },
    {TclDatabase::STATEMENT,	"s"	      }, // shortcut
    {TclDatabase::SET,		strSet	      },
    {TclDatabase::STATEMENT,	strStatement  },
    {TclDatabase::TABLES,	strTables     },
    {TclDatabase::TYPEINFO,	strTypeinfo   }
};

NumStr statementCmds [] = {
    {11,			NULL	    },
    {TclStatement::COLUMNS,	strColumns  },
    {TclStatement::DROP,	strDrop	    },
    {TclStatement::EVAL,	strEval	    },
    {TclStatement::EXECUTE,	strExecute  },
    {TclStatement::FETCH,	strFetch    },
    {TclStatement::GET,		strGet	    },
    {TclStatement::MORERESULTS, strMoreResults },
    {TclStatement::READ,	strRead	    },
    {TclStatement::ROWCOUNT,	strRowCount },
    {TclStatement::RUN,		strRun	    },
    {TclStatement::SET,		strSet	    }
};

#ifdef _DEBUG

typedef struct 
{
    NumStr *ptr_array;
    int		array_size;
} 
NumStrArrayDescriptor;

NumStrArrayDescriptor g_NumStrArrayDescriptors[] = {
    { sqlType,          sizeof( sqlType         ) / sizeof(NumStr)	},
    { attrDef,          sizeof( attrDef         ) / sizeof(NumStr)	},
    { stmtOp,           sizeof( stmtOp          ) / sizeof(NumStr)	},
    { configOp,         sizeof( configOp        ) / sizeof(NumStr)	},
    { connectOp,        sizeof( connectOp       ) / sizeof(NumStr)	},
    { booleanOp,        sizeof( booleanOp       ) / sizeof(NumStr)	},
    { concurrencyOp,    sizeof( concurrencyOp   ) / sizeof(NumStr)	},
    { cursorOp,         sizeof( cursorOp        ) / sizeof(NumStr)	},
    { databaseOptions,  sizeof( databaseOptions ) / sizeof(NumStr)	},
    { databaseCmds,     sizeof( databaseCmds    ) / sizeof(NumStr)	},
    { statementCmds,    sizeof( statementCmds   ) / sizeof(NumStr)	}
};
const unsigned int g_cntNumStrArrays = sizeof(g_NumStrArrayDescriptors)/sizeof(NumStrArrayDescriptor);


// This function ensures that the first entry of each NumStr array indicates the correct number
// of valid entries in the array.  
//
int tclodbc_validateNumStrArrays()
{
    unsigned int claimed_size;
    unsigned int actual_size;
    unsigned int idx;

    NumStr *pArray;

    for (idx=0; idx<g_cntNumStrArrays; ++idx ) {
	pArray = g_NumStrArrayDescriptors[idx].ptr_array;        // reference the array
	claimed_size = pArray[0].numeric;                        // Numeric field of first entry indicates number of remaining valid entries
	actual_size = g_NumStrArrayDescriptors[idx].array_size;  // This includes the first (dummy) entry
	if (actual_size != claimed_size + 1) {
	    return TCL_ERROR;
	}
    }

    return TCL_OK;
}



// Ensure that the StrToNum function gives the correct result for each
// entry in every array.
//
int tclodbc_validateStrToNumFunction()
{
    unsigned int idx, k;
    unsigned int cntValidEntries;
    NumStr *pArray;
    short search_result;

    for (idx=0; idx<g_cntNumStrArrays; ++idx) {
	pArray = g_NumStrArrayDescriptors[idx].ptr_array;
	cntValidEntries = pArray[0].numeric;
	for (k=1; k<=cntValidEntries; ++k) {
	    try {
		search_result = StrToNum(pArray[k].string, pArray,
					 NULL /* errMsg */,
					 FALSE /* allowNumeric */);

		if (pArray[k].numeric != search_result) {
		    return TCL_ERROR;	// breakpoint on this line
		}
	    } catch (...) {
		return TCL_ERROR;	// breakpoint on this line
	    }
	}
    }
    return TCL_OK;	// breakpoint on this line
}

#endif
